full-node/sov-prover-storage-manager/src/lib.rs

```diff
@@ -0,0 +1,411 @@
+                    .entry(prev_block_hash.clone())
+                    .or_default()
+                    .push(current_block_hash.clone());
+
+                self.blocks_to_parent
+                    .insert(current_block_hash, prev_block_hash);
+
+                // Update latest snapshot id
+                self.latest_snapshot_id = new_snapshot_id;
+                new_snapshot_id
+            }
+        };
+
+        let state_db_snapshot = DbSnapshot::new(
+            new_snapshot_id,
+            ReadOnlyLock::new(self.state_snapshot_manager.clone()),
+        );
+
+        let native_db_snapshot = DbSnapshot::new(
+            new_snapshot_id,
+            ReadOnlyLock::new(self.native_snapshot_manager.clone()),
+        );
+
+        Ok(NewProverStorage::with_db_handlers(
+            state_db_snapshot,
+            native_db_snapshot,
+        ))
+    }
+
+    fn save_change_set(&mut self, change_set: Self::NativeChangeSet) -> anyhow::Result<()> {
+        let (state_db, native_db) = change_set.freeze();
+
+        // TODO: Check that this change set belongs to something was issued before
+        {
+            let mut state_manager = self.state_snapshot_manager.write().unwrap();
+            let mut native_manager = self.native_snapshot_manager.write().unwrap();
+
+            state_manager.add_snapshot(state_db.into());
+            native_manager.add_snapshot(native_db.into());
+        }
+
+        Ok(())
+    }
+
+    fn finalize(&mut self, block_header: &Da::BlockHeader) -> anyhow::Result<()> {
+        let current_block_hash = block_header.hash();
+        let prev_block_hash = block_header.prev_hash();
+
+        let snapshot_id = self
+            .block_hash_to_snapshot_id
+            .remove(&current_block_hash)
+            .ok_or(anyhow::anyhow!("Attempt to finalize non existing snapshot"))?;
+
+        let mut state_manager = self.state_snapshot_manager.write().unwrap();
+        let mut native_manager = self.native_snapshot_manager.write().unwrap();
+
+        // Return error here, as underlying database can return error
+        state_manager.commit_snapshot(&snapshot_id)?;
+        native_manager.commit_snapshot(&snapshot_id)?;
+
+        // All siblings of current snapshot
+        let mut to_discard: Vec<_> = self
+            .chain_forks
+            .remove(&prev_block_hash)
+            .expect("Inconsistent chain_forks")
+            .into_iter()
+            .filter(|bh| bh != &current_block_hash)
+            .collect();
+
+        while let Some(block_hash) = to_discard.pop() {
+            let child_block_hashes = self.chain_forks.remove(&block_hash).unwrap_or_default();
+            self.blocks_to_parent.remove(&block_hash).unwrap();
+            let snapshot_id = self.block_hash_to_snapshot_id.remove(&block_hash).unwrap();
+
+            state_manager.discard_snapshot(&snapshot_id);
+            native_manager.discard_snapshot(&snapshot_id);
+
+            to_discard.extend(child_block_hashes);
```

Comment:
Does this loop ends? It looks like we remove one item from `to_discard` and add one on every step of the loop.

Reply:
Yes, it ends, when all "orphans" are removed. Basically it ends, when all heads from non-canonical forks are removed.